atlasdb-dbkvs-tests/src/test/java/com/palantir/atlasdb/keyvalue/dbkvs/DbKvsPostgresInvalidationRunnerTest.java

```diff
@@ -0,0 +1,96 @@
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.dbkvs.impl.ConnectionManagerAwareDbKvs;
+import com.palantir.atlasdb.keyvalue.dbkvs.timestamp.InDbTimestampBoundStore;
+import com.palantir.exception.PalantirSqlException;
+import com.palantir.timestamp.TimestampBoundStore;
+
+public class DbKvsPostgresInvalidationRunnerTest {
+    private final ConnectionManagerAwareDbKvs kvs = DbkvsPostgresTestSuite.createKvs();
+    private final TimestampBoundStore store = getStore();
+    private final InvalidationRunner invalidationRunner = new InvalidationRunner(kvs.getConnectionManager());
+    private static final long TIMESTAMP_1 = 12000;
+
+    @Before
+    public void setUp() throws Exception {
+        kvs.dropTable(AtlasDbConstants.TIMELOCK_TIMESTAMP_TABLE);
+        invalidationRunner.createTableIfDoesNotExist();
+    }
+
+    public InDbTimestampBoundStore getStore() {
+        return InDbTimestampBoundStore.create(
+                kvs.getConnectionManager(),
+                AtlasDbConstants.TIMELOCK_TIMESTAMP_TABLE,
+                DbkvsPostgresTestSuite.getKvsConfig().ddl().tablePrefix());
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        kvs.close();
+    }
+
+    @Test
+    public void poisonsEmptyTableAndReturnsNoOpTs() {
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(NO_OP_FAST_FORWARD_TIMESTAMP);
+    }
+
+    @Test
+    public void poisonsEmptyTableAndReturnsStoredBound() {
+        store.getUpperLimit();
+        store.storeUpperLimit(TIMESTAMP_1);
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void cannotReadAfterBeingPoisoned() {
+        invalidationRunner.getLastAllocatedAndPoison();
+        assertBoundNotReadable();
+    }
+
+    @Test
+    public void poisoningMultipleTimesIsAllowed() {
+        store.storeUpperLimit(TIMESTAMP_1);
+        store.getUpperLimit();
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(TIMESTAMP_1);
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(TIMESTAMP_1);
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(TIMESTAMP_1);
+    }
+
+    @Test
+    public void poisoningEmptyTableMultipleTimesIsAllowed() {
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(NO_OP_FAST_FORWARD_TIMESTAMP);
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(NO_OP_FAST_FORWARD_TIMESTAMP);
+        assertThat(invalidationRunner.getLastAllocatedAndPoison()).isEqualTo(NO_OP_FAST_FORWARD_TIMESTAMP);
+        assertBoundNotReadable();
+    }
+
+    private void assertBoundNotReadable() {
+        assertThatThrownBy(store::getUpperLimit).isInstanceOf(PalantirSqlException.class);
```

Comment:
Currently this keeps retrying for a long time and then fails. So also when we poison the store, readers will behave the same. This is good because there is no risk of corruption, but it would be better to fail more quickly. We can change the read behaviour to allow for that, but in general we cannot succeed in this in case of an atlasdb downgrade.

Reply:
We can check in the store if there is the poison column first, and if yes, then fail immediately